// backend/src/repos/campaignRepo.ts
import type { Knex } from 'knex';

export type CampaignStatus = 'draft' | 'scheduled' | 'sent';

export type CampaignRow = {
  id: number;
  name: string;
  subject: string;
  body: string;
  status: CampaignStatus;
  scheduled_at: Date | null;
  created_by: number;
  created_at: Date;
  updated_at: Date;
};

export async function listCampaignsByUser(db: Knex, userId: number): Promise<CampaignRow[]> {
  return db<CampaignRow>('campaigns')
    .where({ created_by: userId })
    .orderBy('created_at', 'desc');
}

export async function listAllCampaigns(db: Knex): Promise<CampaignRow[]> {
  return db<CampaignRow>('campaigns').orderBy('created_at', 'desc');
}

/** New campaigns start as drafts so content and recipients can be refined before scheduling/sending. */
export async function insertCampaign(
  db: Knex,
  args: { name: string; subject: string; body: string; createdBy: number },
): Promise<CampaignRow> {
  const rows = await db('campaigns')
    .insert({
      name: args.name,
      subject: args.subject,
      body: args.body,
      created_by: args.createdBy,
      status: 'draft',
      scheduled_at: null,
    })
    .returning('*');

  return rows[0];
}

export async function findCampaignByIdForUser(
  db: Knex,
  args: { id: number; userId: number },
): Promise<CampaignRow | null> {
  const row = await db<CampaignRow>('campaigns')
    .where({ id: args.id, created_by: args.userId })
    .first();
  return row ?? null;
}

/** Editing is restricted to drafts; once scheduled/sent, content is treated as immutable for audit/reporting consistency. */
export async function updateDraftCampaign(
  db: Knex,
  args: { id: number; userId: number; patch: Partial<Pick<CampaignRow, 'name' | 'subject' | 'body'>> },
): Promise<CampaignRow | null> {
  const rows = await db('campaigns')
    .where({ id: args.id, created_by: args.userId, status: 'draft' })
    .update(args.patch)
    .returning('*');

  return rows[0] ?? null;
}

/** Only drafts can be deleted to avoid removing historical records for scheduled/sent outreach. */
export async function deleteDraftCampaign(
  db: Knex,
  args: { id: number; userId: number },
): Promise<number> {
  const rows = await db('campaigns')
    .where({ id: args.id, created_by: args.userId, status: 'draft' })
    .delete()
    .returning(['id']);
  return rows.length;
}

/** Scheduling is a reversible pre-send state; sent campaigns cannot be rescheduled. */
export async function scheduleCampaign(
  db: Knex,
  args: { id: number; userId: number; scheduledAt: Date },
): Promise<CampaignRow | null> {
  const rows = await db('campaigns')
    .where({ id: args.id, created_by: args.userId })
    .whereNot({ status: 'sent' })
    .update({ status: 'scheduled', scheduled_at: args.scheduledAt })
    .returning('*');

  return rows[0] ?? null;
}

export async function hasSchedulingConflict(
  db: Knex,
  args: { userId: number; campaignId: number; scheduledAt: Date },
): Promise<boolean> {
  const row = await db<CampaignRow>('campaigns')
    .where({ created_by: args.userId, status: 'scheduled' })
    .whereNot({ id: args.campaignId })
    .where({ scheduled_at: args.scheduledAt })
    .first('id');
  return Boolean(row);
}

/** Marking sent is one-way; downstream analytics assume "sent" never flips back to draft/scheduled. */
export async function markCampaignSent(
  db: Knex,
  args: { id: number; userId: number },
): Promise<CampaignRow | null> {
  const rows = await db('campaigns')
    .where({ id: args.id, created_by: args.userId })
    .whereNot({ status: 'sent' })
    .update({ status: 'sent' })
    .returning('*');

  return rows[0] ?? null;
}
